specs/derivation.md

```diff
@@ -0,0 +1,820 @@
+"zip-bomb" types of attack (where a small compressed input decompresses to a humongous amount of data).
+
+> **TODO** specify `MAX_RLP_BYTES_PER_CHANNEL`
+
+> **TODO** what happens when we reach `MAX_RLP_BYTES_PER_CHANNEL`? do we throw the channel away, or do we still attempt
+> to use the decompressed bytes?
+
+While the above pseudocode implies that all batches are known in advance, it is possible to perform streaming
+compression and decompression of RLP-encoded batches. This means it is possible to start including channel frames in a
+[batcher transaction][g-batcher-transaction] before we know how many batches (and how many frames) the channel will
+contain.bBatch Submission Wire Format
+
+### Batch Format
+
+[batch-format]: #batch-format
+
+Recall that a batch contains a list of transactions to be included in a specific L2 block.
+
+A batch is encoded as `batch_version ++ content`, where `content` depends on the version:
+
+| `batch_version` | `content`                                         |
+| --------------- | ------------------------------------------------- |
+| 0               | `rlp_encode([epoch, timestamp, transaction_list)` |
+
+where:
+
+- `rlp_encode` is a function that encodes a batch according to the [RLP format], and `[x, y, z]` denotes a list
+  containing items `x`, `y` and `z`
+- `epoch` is the [sequencing epoch][g-sequencing-epoch] of the L2 block
+- `timestamp` is the timestamp of the L2 block
+- `transaction_list` is an RLP-encoded list of [EIP-2718] encoded transactions.
+
+[EIP-2718]: https://eips.ethereum.org/EIPS/eip-2718
+
+Unknown versions make the batch invalid (it must be ignored by the rollup node), as do malformed contents.
+
+------------------------------------------------------------------------------------------------------------------------
+
+# Architecture
+
+[architecture]: #architecture
+
+The above describes the general process of L2 chain derivation, and specifies how batches are encoded within [batcher
+transactions][g-batcher-transactions].
+
+However, there remains many details to specify. These are mostly tied to the rollup node architecture for derivation.
+Therefore we present this architecture as a way to specify these details.
+
+A validator that only reads from L1 (and so doesn't interact with the sequencer directly) does not need to be
+implemented in the way presented below. It does however need to derive the same blocks (i.e. it needs to be semantically
+equivalent). We do believe the architecture presented below has many advantages.
+
+## L2 Chain Derivation Pipeline
+
+[pipeline]: #l2-chain-derivation-pipeline
+
+Our architecture decomposes the derivation process into a pipeline made up of the following stages:
+
+1. L1 Traversal
+2. L1 Retrieval
+3. Channel Bank
+4. Batch Decoding (called `ChannelInReader` in the code)
+5. Payload Attributes Derivation (called `BatchQueue` in the code)
+6. Engine Queue
```

Comment:
now there batch queue then payload attributes derivation

Reply:
- batch decoding: decompresses the batches, and send the unordered to the next stage
- batch queue: re-orders the batches and fills in missing batches
     - if contiguous batches available at the start, flush to next stage (eager batch derivation, no need to wait for a full sequencing window)